### PR TITLE
Migrate `x` to using `clap`

### PR DESCRIPTION
**Pull request by clubby789**
cc #107375, #107827, @jyn514 

Beginning experimenting with moving `src/bootstrap/flags.rs` to `clap` instead of `getopts`. A fair chunk of logic is missing but this is basically usable right now, so I'm making a draft PR to have it available.

### CODE REVIEW

**Comment by jyn514** [comment]
cc @Mark-Simulacrum - this adds a few more dependencies, but makes it easier to maintain the arguments and add features like shell completion.

@clubby789 can you time `x --help` before and after this change to see how much of a slow down there is?

**Comment by petrochenkov** [comment]
@jyn514 
> can you time `x --help` before and after this change to see how much of a slow down there is?

If you mean the time including compiling bootstrap including clap, then I'd very much want to see that too.

**Comment by jyn514** [comment]
yes, that's what I meant, with an `x clean` before so it's rebuilt from scratch.

**Comment by clubby789** [comment]
`rm -rf build/bootstrap && time x --help`
master:
`x --help  42.59s user 3.94s system 360% cpu 12.914 total`
x-clap
`x --help  52.26s user 4.70s system 323% cpu 17.625 total`

**Comment by clubby789** [comment]
I'm not sure how to handle the old behaviour of `x <command> -h -v` for listing paths, since `-h` causes the parser to fail. Possibly we could handle the failure, then iterate through the commands looking for a subccomand, `--help/-h` and `--verbose/-{v,vv,vvv}` but that doesn't feel too neat

**Comment by jyn514** [comment]
@clubby789 we could look for -h -v before passing the output to clap; it might not be 100% accurate because of --test-args but in practice I think it will work

File: src/bootstrap/Cargo.toml [hunk 1]
```diff
@@ -34,8 +34,8 @@ build_helper = { path = "../tools/build_helper" }
 cmake = "0.1.38"
 fd-lock = "3.0.8"
 filetime = "0.2"
-getopts = "0.2.19"
 cc = "1.0.69"
+clap = { version = "4.1.4", features = ["derive"] }
```
Comment (klensy) [inline src/bootstrap/Cargo.toml:38]: You can try `default-features = false` and turn on only some of them to reduce build time.

Comment (clubby789) [reply src/bootstrap/Cargo.toml:38]: Looks like it saves a few seconds, thanks

**Comment by clubby789** [comment]
Given that we have clap now, I decided to just have a local `Parser` struct we can use to make sure the parsing logic is the same.
This works enough now (CI runs sucessfully) that I'm going to mark ready for review, but unsure if I'm missing any edge cases/behaviour.

**Comment by albertlarsan68** [comment]
Hello,
I have not yet fully reviewed the PR. Here is the main thing I noticed: Clap has a way to test the correctness of the code (https://docs.rs/clap/latest/clap/_derive/index.html#tips), but I did not see it being used. Can you add it please (as a test)?

**Comment by clubby789** [comment]
I'm working on integrating `clap_complete` for automatic shell completions, but unfortunately it seems a little buggy with subcommands. https://github.com/clap-rs/clap/issues/4652 is one issue I've encountered, as well as zsh erroring due to positional arguments being repeated in all subcommands

File: src/bootstrap/Cargo.toml [hunk 2]
```diff
@@ -34,8 +34,8 @@ build_helper = { path = "../tools/build_helper" }
 cmake = "0.1.38"
 fd-lock = "3.0.8"
 filetime = "0.2"
-getopts = "0.2.19"
 cc = "1.0.69"
+clap = { version = "4.1.4", features = ["std", "usage", "help", "derive", "error-context"], default-features = false}
```
Comment (bjorn3) [inline src/bootstrap/Cargo.toml:38]: How bad would it be to manually implement rather than use the derive macro to reduce the compile time impact?

Comment (jyn514) [reply src/bootstrap/Cargo.toml:38]: I don't think we should focus so heavily on the compile times. If they're an issue, we should finally fix #99989 , which is very close to being done.

**Comment by clubby789** [comment]
Closing in favour of #110693

**Pull request closed without merging**